### PR TITLE
Add busy flag to `Zone`

### PR DESCRIPTION
**Pull request by metaspace**
This patch set prevents a race condition on `Zone::open_for_write_` by adding a busy
flag to the Zone structure. Writing the busy flag through class methods use
acquire/release ordering, while reading the flag uses relaxed ordering. That is, claiming or releasing a zone is a synchronization operation that involves previous and next owners of the zone.

With this patch, users of a zone must acquire ownership of the zone by setting the busy flag of the zone, before managing the zone.

The patch removes `Zone::open_for_write_` as it becomes redundant.

Supersedes #71 
Supersedes #46

### CODE REVIEW

**Comment by metaspace** [comment]
New version to make the linter happy

**Comment by floatious** [comment]
Perhaps give some credit to
https://github.com/westerndigitalcorporation/zenfs/pull/46
by adding a:
Reported-by: Kuankuan Guo guokuankuan@bytedance.com

If it was also reported by Percona, you can add another Reported-by tag
with whoever that found it, so that you have two Reported-by tags.

And maybe overkill, but since Hans suggested that you should use a busy flag,
you could also add a:
Suggested-by: Hans Holmberg <hans.holmberg@wdc.com>

File: fs/io_zenfs.cc [hunk 1]
```diff
@@ -361,7 +361,7 @@ IOStatus ZoneFile::Append(void* data, int data_size, int valid_size) {
   if (!active_zone_) {
     Zone* zone = zbd_->AllocateZone(lifetime_);
     if (!zone) {
-      return IOStatus::NoSpace("Zone allocation failure\n");
+      return IOStatus::NoSpace("Zone allocation failure, no active zone\n");
```
Comment (aravind-wdc) [inline fs/io_zenfs.cc:364]: We could also hit the "No Space" when all the zones are actually full and active zones are less than max-active-zones.  

Would it be better to dump more information such as active-zone-count, full-zone-count, total-zone-count and let the user make sense out of it ?

File: fs/zbd_zenfs.cc [hunk 2]
```diff
@@ -544,20 +579,33 @@ Zone *ZonedBlockDevice::AllocateZone(Env::WriteLifeTimeHint file_lifetime) {
 
     if (active_io_zones_.load() < max_nr_active_io_zones_) {
       for (const auto z : io_zones) {
-        if ((!z->open_for_write_) && z->IsEmpty()) {
-          z->lifetime_ = file_lifetime;
-          allocated_zone = z;
-          active_io_zones_++;
-          new_zone = 1;
-          break;
+        if (z->SetBusy()) {
```
Comment (aravind-wdc) [inline fs/zbd_zenfs.cc:580]: IIRC, SetBusy() would return true if it goes and set busy_ as true, else it returns false.
In case it does set it to true and if the zone is not empty(say it is a closed zone), then we will be setting this zone as busy and unset it in the else part below . 

Wouldn't it be better if we do a if ( ! IsBusy()) here and then actually do SetBusy() when the zone is allocated.

File: fs/fs_zenfs.cc [hunk 3]
```diff
@@ -277,8 +277,8 @@ IOStatus ZenFS::WriteEndRecord(ZenMetaLog* meta_log) {
 
 /* Assumes the files_mtx_ is held */
 IOStatus ZenFS::RollMetaZoneLocked() {
-  ZenMetaLog* new_meta_log;
-  Zone *new_meta_zone, *old_meta_zone;
+  std::unique_ptr<ZenMetaLog> new_meta_log, old_meta_log;
```
Comment (aravind-wdc) [inline fs/fs_zenfs.cc:280]: Do you think it would be better to split the metalog related changes to another patch ?

File: fs/io_zenfs.cc (hunk 1 above)
Comment (metaspace) [reply fs/io_zenfs.cc:364]: This one just says zone allocation failure, nothing about how much space is left. I think caller may add the out of space message. I added this text to distinguish the error from a very similar one in line 384.

Regarding the zone stats, it is available if running with debug. In production mode I think it should be the responsibility of the caller to trigger printing of the stats, if desired.

File: fs/zbd_zenfs.cc (hunk 2 above)
Comment (metaspace) [reply fs/zbd_zenfs.cc:580]: `Zone::IsBusy()` has relaxed memory ordering, so none of the fields of `Zone` is guaranteed to be observed to the value they were set to, before the last owner of the zone called `Zone::UnsetBusy()`. That is to make `Zone::IsBusy()` faster.

We could change `Zone::IsBusy()` to stronger memory ordering, but we would still have a race on `Zone::lifetime_`.

Another approach could be to make a lockless implementation where the `Zone` pointer is atomically updated in-place in the io_zones list with a compare and swap operation. You would have to do this update every time you wanted to change a field of a `Zone`, and you would pass zones around as indexes to the list instead of as pointers. You would still need acquire/release ordering for the compare and swap operation though.

Could be interesting to benchmark the two approaches against each other.

File: fs/fs_zenfs.cc (hunk 3 above)
Comment (metaspace) [reply fs/fs_zenfs.cc:280]: I can do that, I have no opinion on that. But the system would be broken after application of the first patch, before the second one.

File: fs/io_zenfs.cc (hunk 1 above)
Comment (yhr) [reply fs/io_zenfs.cc:364]: I think that "No active zone" adds no value to the end user. "Out of space: zone allocation failure" would be helpful.

File: fs/zbd_zenfs.cc (hunk 2 above)
Comment (yhr) [reply fs/zbd_zenfs.cc:580]: I think we should prioritize that the code should be a) robust and b) easy to understand. The first six rules of optimizations are: don't do it.

As long as Unset/SetBusy is safe to be accessed from different thtreads, i'm good.

File: fs/io_zenfs.cc (hunk 1 above)
Comment (metaspace) [reply fs/io_zenfs.cc:364]: I see that I missed something. Out of space is already added, it is an `IOStatus::NoSpace`. No reason to put it in the message.

I made the change because identical errors were raised in multiple locations. For debugging, it was impossible to know where the error originated. We need either unique error messages, or we need to annotate the error with file name and line number.

File: fs/zbd_zenfs.cc [hunk 4]
```diff
@@ -513,22 +529,41 @@ Zone *ZonedBlockDevice::AllocateZone(Env::WriteLifeTimeHint file_lifetime) {
...
+      if ((z->used_capacity_ > 0) && !z->IsFull()) {
+        unsigned int diff = GetLifeTimeDiff(z->lifetime_, file_lifetime);
+        if (diff <= best_diff) {
+          if (allocated_zone != nullptr) {
+            assert(allocated_zone->UnsetBusy());
+          }
+          allocated_zone = z;
+          best_diff = diff;
+        } else {
+          assert(z->UnsetBusy());
```
Comment (yhr) [inline fs/zbd_zenfs.cc:554]: Danger danger! everything inside assert() will be removed when built DEBUG_LEVEL==0 , not what we want, right?

File: fs/io_zenfs.cc (hunk 1 above)
Comment (yhr) [reply fs/io_zenfs.cc:364]: I think it is motivated to add "out of space" to the message as I recall people reporting "Zone allocation error" to me via email and not understanding that they actually ran out of space..

If we have different allocation errors, we should return what error we hit when returning from AllocateZone

Comment (metaspace) [reply fs/io_zenfs.cc:364]: Printing a `IOStatus::NoSpace` error now gives something like `NoSpace: <message>`. Adding 'out of space' would print `Error ... NoSpace: Zone allocation error, no space`. Do we want that, it is sort of redundant?

> If we have different allocation errors, we should return what error we hit when returning from AllocateZone
That is a good idea, I will add that.

But we still call `AllocateZone` in multiple locations. For debugging, it is desirable to know what call site is propagating the error.

File: fs/zbd_zenfs.cc (hunk 2 above)
Comment (metaspace) [reply fs/zbd_zenfs.cc:580]: > I think we should prioritize that the code should be a) robust and b) easy to understand. 

> As long as Unset/SetBusy is safe to be accessed from different thtreads, i'm good.

I Agree. The current name of the methods does not really convey that you grab a mutex when you call them. How about renaming to `Acquire()` and `Release()` ?

File: fs/zbd_zenfs.cc (hunk 4 above)
Comment (metaspace) [reply fs/zbd_zenfs.cc:554]: Good catch! Will change that. This also reveals that I only ran the test suite with debug mode enabled :confounded:

File: fs/zbd_zenfs.cc (hunk 2 above)
Comment (yhr) [reply fs/zbd_zenfs.cc:580]: Sounds good

File: fs/zbd_zenfs.cc (hunk 4 above)
Comment (MaisenbacherD) [reply fs/zbd_zenfs.cc:554]: This PR is addressing the problem: https://github.com/metaspace/zenfs/pull/1

**Comment by metaspace** [comment]
@yhr Pushed new commits to address the issues mentioned in the comments. I think we can defer adding metaspace/zenfs#1 until we have more bandwidth.

**Review by yhr — APPROVED** [review]
Looks good to me, i'll squash the patches into a single commit and push it manually. @MaisenbacherD   work on top of this can go as a separate pull request to master

**Comment by metaspace** [comment]
Not sure I use the draft/ready feature the right way :man_shrugging:

**Comment by yhr** [comment]
If it's ready for review then it should not be in "draft", so i assumed that this was really in non-draft mode :)

I'm now squashing and fixing up one remaining thing (the pesky this->'s ). I'll make another build & smoke test before pushing

**Comment by yhr** [comment]
Squased & Pushed in 5dd349c0235782785bc472ff7344d9eed924be77
